### DISCUSSION TITLE
TypeLibrary Transformations

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#352] [Ideas]
Pretext;  most COM TypeLibraries are functional but are rarely developer-friendly.  

As a first example, most type libraries do not provide any helpstring documentation, which means that tB cannot currently provide much information in signature-help/intellisense scenarios.

As a second example, in most cases I've come across, the majority of classes are defined in TypeLibraries with the FCANCREATE ('creatable') flag on them, even when the classes are not creatable through COM directly.  

In the Access TypeLibrary for example, all the collections like 'AllForms', 'AllReports', etc, are flagged as creatable in the type library, but you can't create instances of any of them!  They really should NOT have the FCANCREATE flag set on them, as they are meant to be accessed through the object model only.  The real problem with this is that it clutters the intellisense needlessly for the `New <datatype>` dropdown list, and if you did try to create an instance of the class it only fails at runtime, not compile-time.

What I want to see is a text (JSON) format designed to make consistent transformations on a TypeLibrary.  Somthing like this:

```
{
   libraryId: "{guid here}",
   libraryName: "Access", 
   typeInfos: [
      { 
         name: "AllForms",
         type: "TKIND_COCLASS",
         libraryVersion: "*",
         modifications: [
            {
               type: "addDescription",
               value: "This is a collection of forms, returned by the CurrentProject.AllForms property, base index zero..."
            },
            {
               type: "makeNonCreatable"
            },
         ],
      },
      { 
         name: "_AllForms",
         type: "TKIND_INTERFACE",
         libraryVersion: "*",
         modifications: [
            {
               type: "addDescription",
               memberId: "Item",
               memberType: "property-get"
               value: "This is the AllForms::Item property, used for accessing the collection elements by index or by name..."
            },
         ],
      },
   ]
}

```
(above is just a rough idea, the exact format is to be defined...)

We could eventually go further than this, for example by allowing modifications to set hints about allowed datatypes (e.g. if the Item VARIANT property can only accept [String] and [Long], the compiler might issue warnings if other types are given).

Transformations should be considered as HINTS, providing EXTRA detail or information, and not change fundamental ABI compatibility.

The idea of course is that twinBASIC will parse these files and apply the modifications to the internal type library cache to provide much richer intellisense and a better developer experience. 

The plan is that twinBASIC will include the JSON transformation files in its distribution, but a new community-based open source project should be set up to create and maintain the transformation files. 

### EDIT: Versioning

We would want these transformations to be as version-independent as possible, and as such we should avoid referring to types by GUIDs as they generally change with each new version of an interface/class etc.  So instead, we should identify types by their names, just as we would in code.

Implementations that parse the transformations should simply ignore any classes or members that do not match the loaded type library.  This allows for writing the transformations without having to be overly concerned about type library versions.  The transformations are only considered as HINTS to the compiler, so this should be acceptable.

If a transformation really does need to be restricted to a particular library version, then we'd use the libraryVersion field in the JSON to restrict the transformation.  

Thoughts?

**Comment by uwbwsvd** [comment]
My concern is perhaps an own threat.
I don't have the knowledge for a technical opinion.
What I am thinking about here may be something for the distant future, but because I am currently reading a book (The Utopians) I would like to contribute my thoughts here.
Utopian ideas also fail because they have not experienced structure and proper, agreed delegation.
- What can/must be the task of the community? I think that the success of even close-source software depends on the community behind it.
- What advantage does the community have, if tB proves to be extremely profitable, but is also based on the performance of the community?

Translated with www.DeepL.com/Translator (free version)
